src/main/java/codechicken/lib/gui/modular/lib/ScissorHandler.java

```diff
@@ -0,0 +1,81 @@
+package codechicken.lib.gui.modular.lib;
+
+import com.google.common.collect.Queues;
+import com.mojang.blaze3d.platform.Window;
+import com.mojang.blaze3d.systems.RenderSystem;
+import net.minecraft.client.Minecraft;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Deque;
+
+/**
+ * Created by brandon3055 on 29/06/2023
+ */
+public class ScissorHandler {
+    public static final Logger LOGGER = LogManager.getLogger();
+    private final Deque<ScissorState> stack = Queues.newArrayDeque();
+
+    public void pushGuiScissor(double x, double y, double width, double height) {
+        Window window = Minecraft.getInstance().getWindow();
+        int windowHeight = window.getHeight();
+        double scale = window.getGuiScale();
+        double scX = x * scale;
+        double scY = (double) windowHeight - (y + height) * scale;
+        double scW = Math.max(width * scale, 0);
+        double scH = Math.max(height * scale, 0);
+        pushScissor((int) scX, (int) scY, (int) scW, (int) scH);
+    }
+
+    public void pushScissor(int x, int y, int width, int height) {
+        int xMax = x + width;
+        int yMax = y + height;
+        stack.addLast(new ScissorState(x, y, xMax, yMax, stack.peekLast()).apply());
+    }
+
+    public void popScissor() {
+        if (stack.isEmpty()) {
+            LOGGER.error("Scissor stack underflow");
+        }
+        stack.removeLast();
+        ScissorState active = stack.peekLast();
+        if (active != null) {
+            active.apply();
+        } else {
+            RenderSystem.disableScissor();
+        }
+    }
+
+    private static class ScissorState {
```

Comment:
This class can be refactored to a Record, with a static helper method to create it.